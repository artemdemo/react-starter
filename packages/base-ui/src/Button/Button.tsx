import React, { ButtonHTMLAttributes } from 'react';
import classnames from 'classnames';

export enum EButtonAppearance {
  PRIMARY = 'primary',
  SUCCESS = 'success',
  DANGER = 'danger',
  LIGHT = 'light',
}

export enum EButtonSize {
  SM = 'sm',
  LG = 'lg',
}

interface IProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  sizeAppearance?: EButtonSize;
  appearance?: EButtonAppearance;
  block?: boolean;
  'data-testid'?: string;
}

/**
 * Button
 * @link https://getbootstrap.com/docs/4.0/components/buttons/
 */
export const Button: React.FC<IProps> = (props) => {
  const { className, sizeAppearance, block, appearance, ...rest } = props;
  const buttonClass = classnames(className, {
    btn: true,
    'btn-lg': EButtonSize.LG === sizeAppearance,
    'btn-sm': EButtonSize.SM === sizeAppearance,
    'btn-block': block,
    'btn-light': EButtonAppearance.LIGHT === appearance,
    'btn-primary': EButtonAppearance.PRIMARY === appearance || !appearance,
    'btn-danger': EButtonAppearance.DANGER === appearance,
  });
  return (
    <button className={buttonClass} {...rest}>
      {props.children}
    </button>
  );
};
