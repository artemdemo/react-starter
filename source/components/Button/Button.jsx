import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

export const buttonAppearance = {
    PRIMARY: 'primary',
    DANGER: 'danger',
    LIGHT: 'light',
};

export const buttonSize = {
    SM: 'sm',
    LG: 'lg',
};

/**
 * Button
 * @link https://getbootstrap.com/docs/4.0/components/buttons/
 */
const Button = (props) => {
    const { type, className, size, appearance, block, onClick } = props;
    const buttonClass = classnames(className, {
        btn: true,
        'btn-lg': buttonSize.LG === size,
        'btn-sm': buttonSize.SM === size,
        'btn-block': block,
        'btn-light': buttonAppearance.LIGHT === appearance,
        'btn-primary': buttonAppearance.PRIMARY === appearance,
        'btn-danger': buttonAppearance.DANGER === appearance,
    });
    return (
        <button
            className={buttonClass}
            onClick={onClick}
            type={type}
        >
            {props.children}
        </button>
    );
};

Button.propTypes = {
    type: PropTypes.string,
    className: PropTypes.string,
    size: PropTypes.oneOf(Object.values(buttonSize)),
    appearance: PropTypes.oneOf(Object.values(buttonAppearance)),
    block: PropTypes.bool,
    onClick: PropTypes.func,
};

Button.defaultProps = {
    type: 'submit',
    className: '',
    size: undefined,
    appearance: buttonAppearance.PRIMARY,
    block: false,
    onClick: undefined,
};

export default Button;
