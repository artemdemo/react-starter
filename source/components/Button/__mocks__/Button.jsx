import React from 'react';
import _omit from 'lodash/omit';

export const buttonAppearance = {
    PRIMARY: 'primary',
    DANGER: 'danger',
    LIGHT: 'light',
};

export const buttonSize = {
    SM: 'sm',
    LG: 'lg',
};

const Button = (props) => {
    return (
        <div data-mock='Button'>
            {JSON.stringify(_omit(props, 'children'), null, 2)}
            {props.children}
        </div>
    );
};

export default Button;
