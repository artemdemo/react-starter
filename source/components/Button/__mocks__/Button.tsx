import React from 'react';
import _omit from 'lodash/omit';

export const EButtonAppearance = {
    PRIMARY: 'primary',
    DANGER: 'danger',
    LIGHT: 'light',
};

export const EButtonSize = {
    SM: 'sm',
    LG: 'lg',
};

const Button = (props) => {
    return (
        <div data-mock='Button'>
            {JSON.stringify(_omit(props, 'children'), null, 2)}
            {props.children}
        </div>
    );
};

export default Button;
