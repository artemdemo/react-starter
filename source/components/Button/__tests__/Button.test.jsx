/* eslint-disable no-console */
import React from 'react';
import renderer from 'react-test-renderer';
import Button, { buttonAppearance, buttonSize } from '../Button';

describe('<Button>', () => {
    it('Render empty', () => {
        const tree = renderer.create(
            <Button />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('With text', () => {
        const tree = renderer.create(
            <Button>
                Some text
            </Button>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('With type', () => {
        const tree = renderer.create(
            <Button type='button'>
                Some text
            </Button>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('With className', () => {
        const tree = renderer.create(
            <Button className='some-class'>
                Some text
            </Button>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('lg', () => {
        const tree = renderer.create(
            <Button size={buttonSize.LG}>
                Some text
            </Button>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('appearance=light', () => {
        const tree = renderer.create(
            <Button appearance={buttonAppearance.LIGHT}>
                Some text
            </Button>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('appearance=danger', () => {
        const tree = renderer.create(
            <Button appearance={buttonAppearance.DANGER}>
                Some text
            </Button>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('block', () => {
        const tree = renderer.create(
            <Button block>
                Some text
            </Button>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
