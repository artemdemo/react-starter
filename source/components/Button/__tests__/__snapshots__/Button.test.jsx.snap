// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> Render empty 1`] = `
<button
  className="btn btn-primary"
  type="submit"
/>
`;

exports[`<Button> With className 1`] = `
<button
  className="some-class btn btn-primary"
  type="submit"
>
  Some text
</button>
`;

exports[`<Button> With text 1`] = `
<button
  className="btn btn-primary"
  type="submit"
>
  Some text
</button>
`;

exports[`<Button> With type 1`] = `
<button
  className="btn btn-primary"
  type="button"
>
  Some text
</button>
`;

exports[`<Button> appearance=danger 1`] = `
<button
  className="btn btn-danger"
  type="submit"
>
  Some text
</button>
`;

exports[`<Button> appearance=light 1`] = `
<button
  className="btn btn-light"
  type="submit"
>
  Some text
</button>
`;

exports[`<Button> block 1`] = `
<button
  className="btn btn-block btn-primary"
  type="submit"
>
  Some text
</button>
`;

exports[`<Button> lg 1`] = `
<button
  className="btn btn-lg btn-primary"
  type="submit"
>
  Some text
</button>
`;
